internal/controllers/machine/machine_controller_phases.go

```diff
@@ -132,29 +132,26 @@ func (r *Reconciler) reconcileBootstrap(ctx context.Context, s *scope) (ctrl.Res
 	cluster := s.cluster
 	m := s.machine
 
-	// If the Bootstrap ref is nil (and so the machine should use user generated data secret), return.
-	if m.Spec.Bootstrap.ConfigRef == nil {
-		return ctrl.Result{}, nil
-	}
-
-	// Call generic external reconciler if we have an external reference.
-	obj, err := r.reconcileExternal(ctx, cluster, m, m.Spec.Bootstrap.ConfigRef)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			s.bootstrapConfigIsNotFound = true
-
-			if !s.machine.DeletionTimestamp.IsZero() {
-				// Tolerate bootstrap object not found when the machine is being deleted.
-				// TODO: we can also relax this and tolerate the absence of the bootstrap ref way before, e.g. after node ref is set
-				return ctrl.Result{}, nil
+	if m.Spec.Bootstrap.ConfigRef != nil {
```

Comment:
What about 
```suggestion
	// If the bootstrap data secret is set by the user, set ready and return.
	if m.Spec.Bootstrap.ConfigRef == nil {
		if  m.Spec.Bootstrap.DataSecretName == nil {
			return ctrl.Result{}, errors.New("either spec.bootstrap.dataSecretName or spec.bootstrap.configRef must be populated")
		}
		m.Status.BootstrapReady = true
		conditions.MarkTrue(m, clusterv1.BootstrapReadyCondition)
		return ctrl.Result{}, nil
	}
	
	// Call generic external reconciler if we have an external reference.
	...
	
	// Drop >> If the bootstrap data is populated, set ready and return.
	
	...
```
So we take care of user provided data secret first, and then we handle when data secret is generated by the bootstrap provider (without mixing the two use cases)

Reply:
That looks like it would work (though people thought the same about [this](https://github.com/kubernetes-sigs/cluster-api/pull/7394#discussion_r1012924412)). I guess it comes down to whether it's worse to duplicate the code that marks the bootstrap ready, or to duplicate an `if` statement.

Reply:
Feel free to add more tests to ensure the solution works and we don't have regressions.

WRT to the implementation options, I would prefer to mix up the two use cases in the code because the resulting code is simpler to read and to reason about and this usually pays back on the long run.

Reply:
Note for self: Separation between use cases doesn't seems to be addressed in latest commit, also unit test are missing

Reply:
There's obviously no unit tests but the latest commit is exactly what you asked for in terms of separation of the use cases I thought?

Reply:
Sorry but I don't think the last commits address the separation between use cases as the code above will do

Reply:
I'm afraid I'm going to need you to explain in more detail, because the control flow in the updated patch is identical to the snippet you gave above.
It checks for a missing bootstrap ref and handles that case and returns early, leaving the rest of the function to handle the case where there _is_ a bootstrap ref. That's what I understood from when you said:

> take care of user provided data secret first, and then we handle when data secret is generated by the bootstrap provider (without mixing the two use cases)

Reply:
+1 to what Fabrizio wrote above

Reply:
I think the only missing part is that we would prefer inlinig the content of checkMachineBootstrapReady in the two places where it is used because it's easier to read & understand. Then calling a func called `checkMachineBootstrapReady` there

Reply:
To be clear, that was not the trade-off that I was discussing [above](https://github.com/kubernetes-sigs/cluster-api/pull/11459#discussion_r1853236893), though I can see now why probably looks like it when coming along after I made the change, because GitHub is terrible :slightly_frowning_face: 

The original PR looked like [this](https://github.com/kubernetes-sigs/cluster-api/commit/8ebd23574ee9ec5c9b5f6803a6f070c2fcf38cfc) - it rearranged all the code so that there was only a single place where we were marking the bootstrap ready.

Fabrizio (correctly) asked to keep the existing control flow, where the two use cases are separated. I pointed out that it was a trade-off between two types of code duplication, but Fabrizio (correctly) stuck to his guns.

Then I implemented his suggestion, but using a convenience function so that there was in fact no trade-off and we get both the simplified control flow and no duplication.

Subsequently it seems like discussion has proceeded as if the original comments were referring to the new implementation instead of the old one? Which sadly is an occupational hazard of GitHub.

For me the benefit of having `checkMachineBootstrapReady()` is not saving lines, but in having a single implementation of the identical action so it cannot become out of sync (again).